### ISSUE TITLE
Bug: "use strict" should not trigger strict mode in ES3

### ISSUE THREAD

**Issue by fasttime** [eslint/eslint#15456]
### Environment

ESLint version: 8.5.0

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
{
    "parserOptions": {
        "ecmaVersion": 3
    }
}
```
</details>

```js
"use strict";
alert(00);
```

### What did you expect to happen?

Merry Christmas!

No error is expected here. ES3 has no strict mode, the `"use strict"` directive should be ignored, and legacy octal literals like `00` should be always allowed. In fact, the code above works in IE 6...

### What actually happened?

Parser error: invalid number ([demo link](https://eslint.org/demo#eyJ0ZXh0IjoiXCJ1c2Ugc3RyaWN0XCI7XG5cbmFsZXJ0KDAwKTtcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Mywic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==))

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by ljharb** [comment]
It would cause me a lot of problems if this was the default - the current behavior disallows things that work in es3 but break in es5+. Allowing these things with ecma’s version 3 would produce code that can *only ever* work in pre-es5 engines. What’s the usefulness of that?

**Comment by fasttime** [comment]
@ljharb AFAIK, the default language version is ES5 (in case you missed it, I specified `"ecmaVersion": 3` in the configuration). Does this address your concern, or do you think that when the `ecmaVersion` is 3 the parser should only accept code that is valid in **all** ECMAScript versions from 3 upwards?

**Comment by ljharb** [comment]
No, it doesn’t. I explicitly set an ecmaVersion of 3 on 300+ of my npm packages, all of which must work in both es3 browsers and newer ones, and all of which use strict mode.

What is the value of allowing code in ES3 engines that will be guaranteed to break on all newer engines?

**Comment by fasttime** [comment]
@ljharb It's all about correctness. Strict mode was introduced with ES5. Pre-ES5 browsers do not honor strict mode. On the other hand, if I had to forbid certain constructs that are not allowed in strict mode while still keeping ES3 compatibility, I'd rather use rules for that (e.g. [no-octal](https://eslint.org/docs/rules/no-octal), [no-with](https://eslint.org/docs/rules/no-with), etc.). I did not consider the problem of compatibility with existing ESLint configurations, so this would be a breaking change.

**Comment by ljharb** [comment]
If you want sloppy mode semantics, then you wouldn’t include the strict pragma. By including it, you’re already writing es5+ code.

‘It could be done with an option, but the question remains: why is that useful? Sloppy mode semantics are *bad* and to be avoided anyways, and if you need them, you wouldn’t use the pragma.

Are you deploying code to IE 6, and *not* to any ES5 engines? I’d love to understand that use case.

**Comment by fasttime** [comment]
@ljharb Well, if it interests you, I was playing around with some JavaScript compilers written in Java (you know, GraalVM and the Closure Compiler) and I was using ESLint to check my expectations, until I stumbled upon this strange thing with strict mode in ES3. I think of it as an aberration, but the behavior could be intentional as well. I just could not find any bugs or discussions about this topic, so I thought I would open a new issue.

**Comment by ljharb** [comment]
I agree there's some philosophical interest in being able to lint ES3 code and ignore the strict pragma, but as I stated above, I don't think there's any use in that being by default. It could certainly be done via an option, if the eslint folks wanted to maintain that complexity.

**Comment by nzakas** [comment]
Because this is a parsing error and not a linting error, that means it’s coming from Acorn and not ESLint. I’d suggest opening up an issue on Acorn to see what their response is. They usually have a good reason for seemingly incorrect behavior like this that may not be obvious, and if not, they’ll know the best approach to address it.

**Comment by fasttime** [comment]
This was indeed an issue in acorn. It seems to be fixed now: https://github.com/acornjs/acorn/commit/353e24f72bd4b8d4cba3755498bdc862540b4659. Thanks @nzakas!

**Comment by nzakas** [comment]
Awesome. I'm going to reopen because we need to take this into account when we integrate Acorn.

**Comment by btmills** [comment]
We discussed this in the most recent TSC meeting. To summarize that discussion:

- This behavior wasn't intended by ESLint or Acorn. Fixing the bug will produce fewer errors. We treat bug fixes that produce fewer errors as semver-patch changes.
- The recommended way to lint code for both ES3 and ES5 would be to do a matrix lint with both `--parser-options=ecmaVersion:3` and `--parser-options=ecmaVersion:5`.
- To preserve the existing behavior at the parser level, one could fork Acorn and revert the bug fix commit.

I'll remove the breaking label, and we can upgrade Acorn via Espree in a future non-major release.

**Comment by ljharb** [comment]
@btmills so does this mean that all my current packages which are set at ecmaVersion 3, and have files with a strict pragma, will suddenly lose all the strict validation?

It seems like it's worth an option when ecmaVersion is 3 for those that want this checking. "a matrix lint" isn't really practical since it'd be the first time in my experience that linting couldn't be done in one job.

**Comment by fasttime** [comment]
Thanks for the update @btmills. Maybe I'm just overthinking this. The situation I have in mind is that of a file which is currently rejected by the parser and which would become parseable after upgrading Acorn. Currently, such a file would produce one single fatal error in ESLint, while after the upgrade it would start being treated as a valid file, processed by rules, producing zero to many errors (hence potentially more than one). I guess this still counts as "fewer errors" when it comes to versioning ESLint after a bug fix, right? But as I said, probably I'm just being too picky about an edge case like this anyway.

**Comment by nzakas** [comment]
@ljharb yes, that’s correct. It’s unfortunate that you’re relying on this incorrect implementation, but ultimately we want ESLint to be correct in the ways that most users would expect.

If you want to verify that a file works in both ES3 and in strict mode ES5, then it makes sense to run it through ESLint twice, once in each mode. I’m sorry this doesn’t align with your desire to run ESLint just once, but I would think you’d much rather run it twice and get correct results for each run vs a mishmash of results based on a parser bug. 

@fasttime can you give an example of this?

> The situation I have in mind is that of a file which is currently rejected by the parser and which would become parseable after upgrading Acorn. 

I’m just having a hard time imagining what syntax would cause such a situation.

**Comment by ljharb** [comment]
Is it out of the question to add an option on ecmaVersion 3, to restore this behavior that's probably been there forever?

**Comment by fasttime** [comment]
@nzakas Nope, I don't know any real-life example, so maybe that should not be a concern after all.

**Comment by ljharb** [comment]
As for running the file twice - that's never been required in all of eslint's history. If i run it in ecmaVersion 5, it works in every later ecmaVersion, and the same has been true if i run it in ecmaVersion 3, because of this strict mode behavior. This change means that for the first time _ever_ for eslint, you can't set the ecmaVersion to your lowest target, and run eslint once. This feels like a breaking change to me.

**Comment by fasttime** [comment]
@ljharb Why do you think that if you can run a file with `ecmaVersion` 5 you can run it with every later version? Many rules are designed to work differently for different `ecmaVersion`s. Even if you turn off the rules, the parser may produce different AST because of incompatibilities introduced in later versions of the spec (there aren't many, but [some of them](https://webreflection.medium.com/a-tiny-disastrous-ecmascript-change-fadc05c83e69) are fairly notorious). It is even possible for a piece of code to parse without errors with a lower `ecmaVersion` but not with a higher one.

```js
var let = [];
var i = 0;
let[i] = 'foo'; // ok in ES5, error in ES6
```

Don't get me wrong: I would also expect my ES3 code to run as intended in ES5 and later *if* I were implementing today's best practices (using ESLint should be a start). But a `ecmaVersion` one-size-fits-all is not realistic imo.

**Comment by ljharb** [comment]
@fasttime because that's how JavaScript itself works? Overall, syntax is all forwards-compatible, with strict mode being the sole exception (rare edge cases where a library depends on a feature that's removed, but so few websites depend on it that browsers are willing to remove the feature, aren't really relevant).

The reality is that for the entirety of eslint's existence, as well as all the linters it succeeded, this was entirely possible, and eslint has been used in this way. I have never - ever - seen anyone run the linter more than once, on purpose, on the same file, because it's never before been required.

**Comment by fasttime** [comment]
@ljharb You don't need to run ESLint twice if you parse your code with `ecmaVersion` 3 and the rules that replicate strict mode: `no-with`, `no-octal`, `no-dupe-args`, `no-dupe-keys`, `no-delete-var`, `id-denylist` for `eval` and `arguments` and `no-implicit-global`. You might be missing a few edge cases, but no more than you do by assuming current browsers to parse your code with the syntax of ES3 plus strict mode.

**Comment by ljharb** [comment]
I'm pretty confident there aren't any edge cases for that, because in ecmaVersion 3, you can't use `let` or template literals :-)

**Comment by fasttime** [comment]
@ljharb I don't understand... if you have `let` and template literals in your code why do you need `ecmaVersion` 3?

**Comment by ljharb** [comment]
I'm saying I _don't_ - and that without those, there's no edge cases I'm aware of where code will parse as ES3, work in ES3 engines, and NOT work in modern engines.

**Comment by fasttime** [comment]
@ljharb Ah okay. Maybe have a look at my three-line snippet above.

**Comment by ljharb** [comment]
@fasttime that snippet uses the variable name `let`, which no code i've ever seen does, as well as mutates an array (something i never personally do), so it's not a concern for me.

**Comment by nzakas** [comment]
@ljharb I don’t think it makes sense to add an option to allow strict mode directives in ES3 as I don’t think it’s a generally applicable use case. “Parse as ES3 but also with strict mode” isn’t something I can imagine many folks wanting.

I think the best way to enable your current setup would be to continue using the current version of Espree in your setup. The easiest option is probably to create a custom parser that just wraps this version of Espree and use that for all of your runs instead of the default.

I know you’re only running ESLint once right now due to the Acorn bug, but that doesn’t mean it’s the right approach. If you truly want both ES3 and ES5 compatibility, with support for strict mode, and you want those runs to be 100% valid, the only way to do that is to run it twice.

Again, I truly am sorry this will inconvenience you, but it’s important for ESLint to behave as correctly as possible and sometimes that means fixing longstanding bugs that will result in different behavior.

**Comment by ljharb** [comment]
I think my use case being automatically for over 10% of npm's download traffic should qualify it as generally acceptable. The reality is that making this change across 300+ repos is simply untenable without an option i can put in a shared config, which means there'll just be a risk of bugs, and that really sucks for the ecosystem :-/

Bug or not, is there any indication it hasn't been present in all earlier incarnations of eslint? Meaning, isn't there a pretty high risk of breakage here to make such a change in a non-major?

**Comment by nzakas** [comment]
We’ve already decided on the course of action here, and we know that this will inconvenience you, but again, this is a serious bug that we can’t let go or wait for a major release. Acorn released this as a patch fix so we can’t get other updates without also getting this one. 

Im assuming you must have a way to deal with other breaking changes across all of your repos? What about just pinning the ESLint version?

**Comment by ljharb** [comment]
It won't immediately break all of my repos, of course; it's more that I have no recourse after the fact.

I don't agree it's a serious bug - it's merely technically incorrect, but "linting ES3 code with a strict pragma that would throw in ES5+" is simply _not_ a use case anyone has.

**Comment by ljharb** [comment]
@nzakas is there any chance the eslint team would be open to a parser flag to restore this behavior? i so i'd be happy to file an RFC; it will really suck for all my packages to be permanently stuck on eslint v8.8.

**Comment by nobiebonie** [comment]
yes

**Comment by nzakas** [comment]
@ljharb no, for all the reasons already discussed.

**Comment by ljharb** [comment]
@nzakas i still don't understand why this is a "serious bug" - the "buggy" behavior - which every linter, eslint or otherwise, has always done until now, afaict - would only have broken someone who was shipping to IE 6-8 (or older ES3 browsers) and NOT shipping to anything that supports strict mode. I really don't understand why that's an audience that needs to be catered to.

**Comment by aladdin-add** [comment]
@ljharb the bug has been fixed in acorn v8.7.1 (patch release), so we won't be able to keep the old behaviour unless pin acorn v8.7.0 in espree. :)

**Comment by ljharb** [comment]
@aladdin-add yes, and i disagree that it’s a bug. It’s a critical feature, and this “bugfix” has been a breaking change for me, and now i have 300+ projects pinned to eslint 8.8.0 as a result.
